apps/dgst.c

```diff
@@ -571,8 +554,7 @@ int do_fp(BIO *out, unsigned char *buf, BIO *bp, int sep, int binout, int xoflen
 
         BIO_get_md_ctx(bp, &ctx);
         if (!EVP_DigestSignFinal(ctx, NULL, &tmplen)) {
-            BIO_printf(bio_err, "Error Signing Data\n");
-            ERR_print_errors(bio_err);
+            BIO_printf(bio_err, "Error getting length of signed data\n");
```

Comment:
Is this the only failure mode?

Reply:
AKAICS yes because `EVP_DigestSignFinal(ctx, NULL, &tmplen)` just asks for the size.
The `EVP_DigestSignFinal(ctx, buf, &len)` below returns the signature, 
and I believe the actual signature calculation is done along with that.
Yet `EVP_DigestSignInit.pod` is not entirely clear to me on this.

Reply:
Just verified my suspicion on `dsa_sign()` and `ecdsa_sign()`, so: yes.

Reply:
I'm proposing a respective doc improvement in #15354.

Reply:
Now add a 3rd party provider into the mix.  Is this still the only failure mode?

The message needs to less specific.

Reply:
A 3rd party provider should behave according to the expectations that hopefully are sufficiently documented. My above reasoning was just to convince myself of what those expectations should be, and doing so I even updated the documentation what the users of `EVP_DigestSignFinal()` can expect.

The response meanwhile given below by @t8m confirms that for the caller of `EVP_DigestSignFinal(ctx, NULL, &tmplen)` the point is just that the `dgst` app obtains the (maximum) length of the signed data, and the error message should refer to that. 
I've made the error message more precise in this way: `"Error getting maximum length of signed data\n"`
And IMHO this is all what can be said reasonably in this error message.

Note that also in the call further down `EVP_DigestSignFinal(ctx, buf, &len)`
everyone is happy with `"Error signing data\n"`, although theoretically a provider could in addition do any other/further stuff like mining bitcoins, which also could result in a failure of the function that would be unrelated to signing the data, but of course we do not take such things into account in the error message.

I don't think it's worth spending much effort in discussing such error messages at length and in detail.
My original aim when replacing `"Error signing data\n"` was just to correct that statement because it was plainly wrong/misleading, and this correction is certainly achieved now with the new formulation.

Reply:
@paulidale, is this okay now?